### ISSUE TITLE
gpui macOS: blurry background broken shadow

### ISSUE THREAD

**Issue by Congyuwang** [zed-industries/zed#10993]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

On macOS, when the window background is set to blurry. Sometimes the shadow flashes, and the shadow can appear only partially, which looks broken.

<img width="365" alt="截圖 2024-04-25 22 48 28" src="https://github.com/zed-industries/zed/assets/52687642/2d1c6eb1-0d10-46f4-8318-b6303220bfab">

### Environment

Zed: v0.133.2 (Zed Preview)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: aarch64

**Comment by Congyuwang** [comment]
Also has weird thumbnail rendering when swiped up to show all desktops:

- ghost rectangles on top left:
![截圖 2024-04-26 05 38 15](https://github.com/zed-industries/zed/assets/52687642/14c772c7-db2f-4feb-9ebb-d59762d18cd6)

![截圖 2024-04-26 05 38 33](https://github.com/zed-industries/zed/assets/52687642/0134e1d8-eac1-4813-a0ab-3cf320f757e0)

- sometimes no ghost, but it should look blurry. It looks transparent instead (as compared to Safari on the left).
![截圖 2024-04-26 05 39 54](https://github.com/zed-industries/zed/assets/52687642/3e1f7021-ea1a-4aff-996d-26e623e8e268)

**Comment by jansol** [comment]
The ghosts look exactly like ghost text that sometimes appears on transparent windows... Which turns out to be that the non-transparent parts of the window are actually holes in the shadow and the shadow does not get updated unless manually triggered or the window resizes. However updating the shadows on every frame was found in Alacritty to be very slow, so the next best workaround is to disable shadows entirely.

I can repro the non-blurred transparent thumbnail window on macOS 12.7.4 as well. I guess just add it to the list of known problems with `CGSSetWindowBackgroundBlurRadius` for now.

**Comment by Congyuwang** [comment]
Hi, @jansol, thanks for your reply. I've tried NSVisualEffectView in an empty Xcode Project (not yet in Zed). It certainly can be set as light so as not to follow system appearance:
<img width="1120" alt="截圖 2024-04-26 17 19 19" src="https://github.com/zed-industries/zed/assets/52687642/c9417427-22c1-4c23-b59f-3eae296d6903">

https://github.com/zed-industries/zed/assets/52687642/f18aefd9-5bab-4612-bb0b-993200cece98

I'm not very familiar with Mac app development though. Do you think this might work for blurry background case, or is there still some problems? (i.e., set background to this `Appearance: Vibrant Light` + `Material: Under Window Background`, and above which applies Zed themes with their different color tints)?

I'm willing to spend some more time figuring out how to put it into Zed (no idea how to do that just yet), if it could in theory work.

**Comment by jansol** [comment]
Vibrant Light adds a 80% opaque white layer which *is* the problem I mentioned earlier. (And Vibrant Dark adds a similar black one, both are a problem for themes)

For themes to look consistent, we need to make that "backdrop" layer completely transparent (or remove it), since Zed draws its own backdrop on top of the transparent background.

**Comment by Congyuwang** [comment]
Okay, I see. But that involves messing with private layers which might change over macOS versions. So, there is not really a neat solution.

Also, it is interesting that macOS Terminal App doesn't have any ghost image or shadow problem, although it also uses `CGSSetWindowBackgroundBlurRadius`. Perhaps it is because that terminal is simpler and it has only very few "layers"? The ghost and shadow problem seems particularly apparent when Mission Control shrinks the window, and the shadow just keeps flashing.

Even though, I still feel the window shadow quite important to blurry light themes. Most Mac app windows do have shadows too, which make them stand out.

**Comment by Congyuwang** [comment]
Perhaps the problem isn't about `CGSSetWindowBackgroundBlurRadius`. After removing its usage completely, the ghost shape and broken shadow still appear in transparent background.

On the other hand, macOS Terminal App doesn't have any ghost image or shadow problem, although it also uses `CGSSetWindowBackgroundBlurRadius`.

**Comment by jansol** [comment]
Yes, the problems are not directly related to `CGSSetWindowBackgroundBlurRadius` but to the window shadow, which is why I opened this PR to disable shadows until someone figures out a better solution.

**Comment by Congyuwang** [comment]
Update. For blurry window, the shadow problem seems to appear only when **_Stage Manager_** is on, see https://github.com/alacritty/alacritty/issues/6229#issuecomment-1198950884. And in my own testing it really is so. So, I am thinking that, perhaps the fix can turn off shadow only for transparent window?

**Comment by Congyuwang** [comment]
Related issue in Alacritty: https://github.com/alacritty/alacritty/issues/6229

**Comment by hfjallemark** [comment]
I can still reproduce it in `Zed Preview 0.159.2`.

**Comment by Aaron-212** [comment]
I wonder if it is possible to use NSVisualEffectView rather than calling private APIs?

**Comment by Congyuwang** [comment]
> Vibrant Light adds a 80% opaque white layer which _is_ the problem I mentioned earlier. (And Vibrant Dark adds a similar black one, both are a problem for themes)
> 
> For themes to look consistent, we need to make that "backdrop" layer completely transparent (or remove it), since Zed draws its own backdrop on top of the transparent background.

I think NSVisualEffectView is worth trying. Although as Jansol  mentioned, NSVisualEffect not only does blurriness, but also applies opaqueness and white / dark color to it.

I am of the opinion that themes for blurry background should be specialized for blurry background, and should adapt to NSVisualEffect’s opaqueness (build on top of that).

If a theme is designed not for blurry background, then don’t use it with blurry background.

Now I never use blurry backgrounds because of the lack of window shadow.

**Comment by hfjallemark** [comment]
I can still reproduce in `0.178.3`
